### ISSUE TITLE
Request programming language support here

### ISSUE THREAD

**Issue by rien** [dodona-edu/dolos#1029]
If you want to use a programming language that Dolos does not support yet, please ask here! This helps us with prioritizing which programming languages we should focus on first.

We currently ship Dolos with the following programming languages:
- Bash
- C
- C++
- C#
- Elm
- Java
- JavaScript
- Python
- Typescript
- Tsx

If your programming language is not in the list of languages supported out-of-the box, there is a high possibility that a tree-sitter parser already exists for that language. If that is the case, it should be easy to add support for your language.

In any case, let us know which languages you want to use with Dolos!

**Comment by BTWS2** [comment]
Adding support for HTML would be great, because the [HTML judge](https://github.com/dodona-edu/judge-html) allows for exercises with different solutions (e.g. add a title (doesn't matter which text), add at least 3 items to a list, ...).

**Comment by rien** [comment]
@BTWS2 there exists a parser for HTML, but it would not for plagiarism detection because the parser ignores tag names, attribute names, the exact content itself, ... 

This is how tree-sitter converts an example HTML page:
![image](https://user-images.githubusercontent.com/3226995/214796875-4563ad9f-3027-4b4f-b4d3-c779c4631dc5.png)

Especially if the underlying structure of the analyzed HTML files is expected to be very similar, using this parser would result in very high similarities. Using this parser, Dolos reports that the homepage of GitHub and the homepage of Dodona have a similarity of 88%.

I would prefer to stick to languages that work good with Dolos. However if you want, you can try it out yourself by installing `tree-sitter-html` using npm or yarn, Dolos is able to automatically detect and use this parser if it is available.

**Comment by BTWS2** [comment]
@rien No problem, thank you for the insight.

**Comment by anilgulecha** [comment]
@rien can there support simple text?

Regular assignments (like essays) is the usecase I'm thinking for this. The tokenizer is as simple as splitting by space?

**Comment by rien** [comment]
@anilgulecha Dolos is specifically made for plagiarism detection on source code. There are tools that should perform better on just text than Dolos.

That said, we do indeed have a tokenizer that does split on spaces which you can use by passing `--language char`. However, in that case you might be better / faster using the `diff` command or something else that does string matching.

**Comment by yafuerst** [comment]
I need support for the language Modelica. It is not supported by tree-sitter, but there are two parsers on github:
https://github.com/OpenModelica/tree-sitter-modelica
https://github.com/mtiller/modelica-tree-sitter

I managed to get them running using tree-parser directly, but I had no luck adding them to dolos yet. Do you have any tips? I am on Linux if thats important.

**Comment by rien** [comment]
@yafuerst Dolos will try to find the parser with a fitting name (`tree-sitter-${name}`) if you add the language with the `-l` option. It will look in the `node_modules` accessible to Dolos (local, per user, global).

If you've managed to get them working but if Dolos doesn't work, you can try "installing" the parser fro your user or globally with `npm link` or `npm link --global`. For the `modelica-tree-sitter` parser to be detected by Dolos, you will have to change the name to `tree-sitter-modelica`.

Let me know if it doesn't work and we'll figure it out.

**Comment by rien** [comment]
@alexey-sh since those languages use multiple languages (template syntax, css, html, ...), tree-sitter does not handle those out-of-the box, so some additional work is required for them to work.

In addition, since HTML and CSS often have a lot of common code fragments between submissions, Dolos isn't very good in detecting plagiarism with them (you get a lot of false positives).

However, we do plan on changing Dolos under the hood to support these kind of languages in the future!

**Comment by DhruvDh** [comment]
Hi, I am running dolos with the following version -

```
Dolos v2.3.0
Node v18.16.0
Tree-sitter v0.20.1
```

`npm` only has `tree-sitter-java@0.19.1`, and it seems `dolos` cannot find it because of this. Any workaround? I have tried installing it locally and globally with `pnpm` and `npm`.

**Comment by rien** [comment]
@DhruvDh with the way we currently integrate tree-sitter languages, we will have to wait on `tree-sitter-java` to publish a new release. Recently, someone already made an issue with the maintainers of that parser to create a new release, let's hope they publish it soon: https://github.com/tree-sitter/tree-sitter-java/issues/163

As an alternative, you can try cloning this repository and updating the base tree-sitter version manually. However, that van be cumbersome.

We already have some ideas how to avoid this problem with Dolos in the future (see #1028), however we've not started on the implementation of this solution yet.

**Comment by DhruvDh** [comment]
I was able to solve this by the following, I am not confident I understand it correctly, so I won't attempt an explanation.
```
# a fork with package.json version set to 0.20.1
pnpm install DhruvDh/tree-sitter-java 
pnpm rebuild
pnpm install @dodona/dolos
pnpm exec dolos run info.csv
```

**Comment by nachiket** [comment]
Would you support Verilog? Thanks.

**Comment by rien** [comment]
Hi @nachiket, thanks for your suggestion! There is an official verilog parser available, so this is definitely possible.

We'll put it in our schedule and will let you know when support for verilog has landed.

**Comment by rhz1949** [comment]
Would you support Go and Rust? Thanks

**Comment by rien** [comment]
@rhz1949 working on that right now. If everything goes well, support for Go and Rust will be added in the new release I will make today.

**Comment by kudhru** [comment]
Is it posible to support Ocaml as well?

**Comment by rien** [comment]
@kudhru yes, there seems to be an official Ocaml parser (https://github.com/tree-sitter/tree-sitter-ocaml).

It seems they provide three parsers: for implementations (.ml) interfaces (.mli) and types. Will it work for you if we only support the implementations? Because Dolos currently supports single-file submissions anyway.

**Comment by kudhru** [comment]
Yes, in the programming exam I am conducting, I will have only one `.ml` file submission only per student.

But how much time would it take? Or could you tell me if I can do it locally on my machine? As I need to conduct an end-semester exam after 10 days.

**Comment by benhadid** [comment]
hello, 
is it possible to support mips / risc-v assembly languages
thank you

**Comment by rien** [comment]
@benhadid I cannot finds a specific parser for MIPS / RISC-V, but there seems to be a generic [tree-sitter-asm](https://github.com/RubixDev/tree-sitter-asm/) parser, can you check whether this would fit your use-case? You can do so by checking their test cases: https://github.com/RubixDev/tree-sitter-asm/tree/main/test/corpus

If not, is there any other parser from [tree-sitter's list of parsers](https://github.com/tree-sitter/tree-sitter/wiki/List-of-parsers) that would fit?

**Comment by benhadid** [comment]
@rien I checked the links you gave me and unfortunately this parser [tree-sitter-asm](https://github.com/RubixDev/tree-sitter-asm/) is for the x86 architecture (intel/amd processors).

I have googled a bit and found this [tree-sitter-riscasm](https://github.com/erihsu/tree-sitter-riscvasm) parser for RISC-V. Can it be used with Dolos ? 

Thank you
